common/utils/include/claragenomics/utils/allocator.hpp

```diff
@@ -225,23 +223,57 @@ class CachingDeviceAllocator
     /// @param stream CUDA stream to be associated with this method.
     void deallocate(pointer p, std::size_t n, cudaStream_t stream = 0)
     {
-        CGA_CU_ABORT_ON_ERR(cub_allocator_->DeviceFree(p));
+        // deallocate should not throw execeptions which is why CGA_CU_CHECK_ERR is not used.
+        CGA_CU_ABORT_ON_ERR(memory_resource_->DeviceFree(p));
     }
 
-    /// @brief returns a shared pointer to internally used cub::CachingDeviceAllocator
-    /// @return a shared pointer to internally used cub::CachingDeviceAllocator
-    std::shared_ptr<cub::CachingDeviceAllocator> cub_allocator() const { return cub_allocator_; }
+    /// @brief returns a shared pointer to memory_resource
+    /// @return a shared pointer to memory_resource
+    std::shared_ptr<MemoryResource> memory_resource() const { return memory_resource_; }
 
 private:
-    std::shared_ptr<cub::CachingDeviceAllocator> cub_allocator_;
+    /// @brief Special constructor for cub::CachingDeviceAllocator
+    /// @param max_cached_bytes
+    /// @tparam is_cub_allocator true
+    /// @return memory resource
+    template <bool is_cub_allocator = std::is_same<MemoryResource, cub::CachingDeviceAllocator>::value>
+    std::enable_if_t<is_cub_allocator, std::shared_ptr<MemoryResource>>
+    generate_memory_resource(size_t max_cached_bytes)
+    {
+        /// Smallest cached bin is 2^10 bytes, largest is 2^28 bytes. All allocation requests larger than 2^28 bytes are not fit in a bin and are not cached
+        return std::make_shared<cub::CachingDeviceAllocator>(2, 10, 28, max_cached_bytes, false, false);
+    }
+
+    /// @brief Constructor for all other memory resources
+    /// @param max_cached_bytes
+    /// @tparam is_cub_allocator false
+    /// @return memory resource
+    template <bool is_cub_allocator = std::is_same<MemoryResource, cub::CachingDeviceAllocator>::value>
+    std::enable_if_t<!is_cub_allocator, std::shared_ptr<MemoryResource>>
+    generate_memory_resource(size_t max_cached_bytes)
+    {
+        return std::make_shared<MemoryResource>(max_cached_bytes);
+    }
+
+    std::shared_ptr<MemoryResource> memory_resource_;
 };
 
-#ifdef CGA_ENABLE_ALLOCATOR
-/// Default device allocator do be used if CGA_ENABLE_ALLOCATOR is set
-using DefaultDeviceAllocator = CachingDeviceAllocator<char>;
+#ifdef CGA_ENABLE_CACHING_ALLOCATOR
+
+#ifdef CGA_ENABLE_PREALLOCATING_ALLOCATOR
+using DefaultDeviceAllocator = CachingDeviceAllocator<char, DevicePreallocatedAllocator>;
+#else
+using DefaultDeviceAllocator = CachingDeviceAllocator<char, cub::CachingDeviceAllocator>;
+#endif
+
+#else
+
+#ifdef CGA_ENABLE_PREALLOCATING_ALLOCATOR
+static_assert(false, "Preallocating allocator can only be used together with caching allocator");
```

Comment:
Surprising enough it did work... kind of:
```
[...]/ClaraGenomicsAnalysis/common/utils/include/claragenomics/utils/allocator.hpp(268): error: static assertion failed with "Preallocating allocator can only be used together with caching allocator"
[...]/ClaraGenomicsAnalysis/common/utils/include/claragenomics/utils/device_buffer.hpp(31): error: expected an identifier
[...]/ClaraGenomicsAnalysis/common/utils/include/claragenomics/utils/device_buffer.hpp(31): error: expected a type specifier
[...]/ClaraGenomicsAnalysis/common/utils/include/claragenomics/utils/device_buffer.hpp(31): error: expected a type specifier
[...]
```

Not a perfect error, but still better than putting some random text to simply make the compilation fail. What do you think?

Reply:
How about a more traditional `#error` with the preprocessor? Haven't used it in a long time, but I think it should be a good fit here.
Since this is an error a user of the library is likely to see based on their choice of cmake options, it should be as clear as possible.

Reply:
good point, done